WcaOnRails/app/controllers/teams_controller.rb

```diff
@@ -1,53 +1,69 @@
 class TeamsController < ApplicationController
   before_action :authenticate_user!
-  before_action -> { redirect_unless_user(:can_manage_teams?) }, except: [:edit, :update]
-  before_action -> { redirect_unless_user(:can_edit_team?, team_from_params) }, only: [:edit, :update]
-
-  def index
-    @teams = Team.all
-  end
+  before_action -> { redirect_unless_user(:can_manage_committees?) }, only: [:new, :create, :destroy]
+  before_action -> { redirect_unless_user(:can_edit_team?, team_from_params_slug) }, only: [:edit]
+  before_action -> { redirect_unless_user(:can_edit_team?, team_from_params) }, only: [:update]
 
   def new
-    @team = Team.new
+    @committee = Committee.find_by_slug(params[:committee_id])
+    @team = @committee.teams.new
   end
 
   def create
-    @team = Team.new(team_params)
+    @committee = Committee.find(params[:committee_id])
+    @team = @committee.teams.new(team_params)
     if @team.save
-      flash[:success] = "Created new team"
-      redirect_to edit_team_path(@team)
+      flash[:success] = "Created new team:" + @team.name
+      redirect_to committee_path(@committee.slug)
     else
-      flash[:danger] = "Failed to create team"
+      flash[:danger] = "Failed to create team:" + @team.name
       render :new
     end
   end
 
   def edit
-    @team = team_from_params
+    @committee = Committee.find_by_slug(params[:committee_id])
+    @team = team_from_params_slug
   end
 
   def update
+    @committee = Committee.find(params[:committee_id])
     @team = team_from_params
     if @team.update_attributes(team_params)
       flash[:success] = "Updated team"
-      redirect_to edit_team_path(@team)
+      redirect_to committee_path(@committee.slug)
     else
       flash.now[:danger] = "Failed to update team"
       render :edit
     end
   end
 
-  private def team_params
-    team_params = params.require(:team).permit(:name, :description, :friendly_id, team_members_attributes: [:id, :team_id, :user_id, :start_date, :end_date, :team_leader, :_destroy])
-    if team_params[:team_members_attributes]
-      team_params[:team_members_attributes].each do |member|
-        member.second.merge!(current_user: current_user.id)
+  def destroy
+    @committee = Committee.find(params[:committee_id])
+    @team = team_from_params
+    if @team.team_members.empty?
+      if @team.destroy
+        flash[:success] = "Successfully deleted team!"
+        redirect_to committee_path(@committee.slug)
+      else
+        flash[:error] = "Error: Could not delete team" + @team.name
+        redirect_to committee_path(@committee.slug)
       end
+    else
+      flash[:error] = "Cannot delete a team whilst it still has team members. If you really want to delete this team then delete the team members first."
+      redirect_to committee_path(@committee.slug)
```

Comment:
In each case you perform the same redirection, so it can be moved out of the `if` statement.